src/utility/NumericalDerivatives.cpp

```diff
@@ -0,0 +1,115 @@
+#include <BSMPT/utility/NumericalDerivatives.h>
+
+namespace BSMPT
+{
+
+std::vector<double>
+NablaNumerical(const std::vector<double> &phi,
+               const std::function<double(std::vector<double>)> &f,
+               const double &eps,
+               const int &dim)
+{
+  std::vector<double> result(dim);
+
+  for (int i = 0; i < dim; i++)
+  {
+    std::vector<double> lp2 = phi;
+    lp2[i] += 2 * eps;
+    std::vector<double> lp1 = phi;
+    lp1[i] += eps;
+    std::vector<double> lm1 = phi;
+    lm1[i] -= eps;
+    std::vector<double> lm2 = phi;
+    lm2[i] -= 2 * eps;
+    result[i] = (-f(lp2) + 8 * f(lp1) - 8 * f(lm1) + f(lm2)) / (12 * eps);
+  }
+  return result;
+}
+
+
+std::vector<std::vector<double>> HessianNumerical_BI(
+    const std::vector<double> &phi,
+    const std::function<double(std::vector<double>)> &V,
+    double eps,
+    const int &dim)
+{
+  std::vector<std::vector<double>> result(dim, std::vector<double>(dim));
+  for (int i = 0; i < dim; i++)
+  {
+    // https://en.wikipedia.org/wiki/Finite_difference
+    for (int j = i; j < dim; j++)
+    {
+      double r = 0;
+
+      if (i == j) eps /= 2;
+
+      std::vector<double> xp = phi; // F(x+h, y+h)
+      xp[i] += eps;
+      xp[j] += eps;
+      r += V(xp);
+
+      xp = phi; //-F(x+h, y-h)
+      xp[i] += eps;
+      xp[j] -= eps;
+      r -= V(xp);
+
+      xp = phi; //-F(x-h, y+h)
+      xp[i] -= eps;
+      xp[j] += eps;
+      r -= V(xp);
+
+      xp = phi; // F(x-h, y-h)
+      xp[i] -= eps;
+      xp[j] -= eps;
+      r += V(xp);
+
+      result[i][j] = r / (4 * eps * eps);
+      result[j][i] = r / (4 * eps * eps);
+
+      if (i == j) eps *= 2;
```

Comment:
I get the idea of halving the epsilon because you would add 2 epsilon to x_i otherwise.
But does the `epsilon /=2` here not remove the factor 4? 
So should this be turned in a different order?
```suggestion
      if (i == j) eps *= 2;
      result[i][j] = r / (4 * eps * eps);
      result[j][i] = r / (4 * eps * eps);


```

Reply:
I think the order is correct. We should not change step size in the middle of the calculation.
From [Wikipedia](https://en.wikipedia.org/wiki/Finite_difference#Rules_for_calculus_of_finite_difference_operators) the expression for the double $x$ derivative is

$f_{x x}(x,y)\approx{\frac{f(x+h,y)-2f(x,y)+f(x-h,y)}{h^{2}}}$

and for the $x,y$ derivative is

$f_{x y}(x,y)\approx{\frac{f(x+h,y+k)-f(x+h,y-k)-f(x-h,y+k)+f(x-h,y-k)}{4h k}}$.

If $x=y$ and I don't half the step size then I will calculate a derivative with twice the step size. (By halving it indeed removes the $4$ and matches the expression above).

I have two implementations for the Hessian (which should be corrected) but the difference is only on the diagonal.

Reply:
Ok but then something else is causing the tests to fail if I use the BI Hessian in the minimum_tracer file

Reply:
Just to brainstorm: Could it be that the reference in the unit test is wrong? From what you say the BI method seems to be the correct one but leads to failing unit tests.

Reply:
I don't think that unit test is wrong, we just check for the number of phases, and we clearly see some unstable behaviour.
I am not sure why this is happening. **Not** halving the ```eps``` produces stabler results and I actually think this is the correct one since the error is of the same order as the other ones.
I also noticed that increasing the constant that we add to the diagonals of the Hessian matrix in the LocateMinimum produces stabler results, but it does not completely solve the problem. The problem seems to be coming from the Hessian not being very stable, one more reason to implement analytical derivatives.

I think we should use the version that does not halve the step (all tests pass this way) and increase the diagonal constant as a precaution (the tests still pass and computational time is not altered).